xapian-core/cluster/cluster.cc

```diff
@@ -0,0 +1,586 @@
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+
+#include <config.h>
+
+#include "xapian/cluster.h"
+
+#include "cluster/clusterinternal.h"
+#include "xapian/error.h"
+#include "api/termlist.h"
+#include "debuglog.h"
+#include "omassert.h"
+
+#include <cmath>
+#include <unordered_map>
+#include <vector>
+
+using namespace Xapian;
+using namespace std;
+
+FreqSource::~FreqSource()
+{
+    LOGCALL_DTOR(API, "FreqSource");
+}
+
+Similarity::~Similarity()
+{
+    LOGCALL_DTOR(API, "Similarity");
+}
+
+Clusterer::~Clusterer()
+{
+    LOGCALL_DTOR(API, "Clusterer");
+}
+
+TermListGroup::TermListGroup(const MSet &docs)
+{
+    LOGCALL_CTOR(API, "TermListGroup", docs);
+    for (MSetIterator it = docs.begin(); it != docs.end(); ++it)
+	add_document(it.get_document());
+    docs_num = docs.size();
+}
+
+doccount
+DummyFreqSource::get_termfreq(const string &) const
+{
+    LOGCALL(API, doccount, "DummyFreqSource::get_termfreq", NO_ARGS);
+    return 1;
+}
+
+doccount
+DummyFreqSource::get_doccount() const
+{
+    LOGCALL(API, doccount, "DummyFreqSource::get_doccount", NO_ARGS);
+    return 1;
+}
+
+void
+TermListGroup::add_document(const Document &document)
+{
+    LOGCALL_VOID(API, "TermListGroup::add_document", document);
+
+    TermIterator titer(document.termlist_begin());
+    TermIterator end(document.termlist_end());
+
+    for (; titer != end; ++titer) {
+	unordered_map<string, doccount>::iterator i;
+	i = termfreq.find(*titer);
+	if (i == termfreq.end())
+	    termfreq[*titer] = 1;
+	else
+	    i->second += 1;
+    }
+}
+
+doccount
+TermListGroup::get_doccount() const
+{
+    LOGCALL(API, doccount, "TermListGroup::get_doccount", NO_ARGS);
+    return docs_num;
+}
+
+doccount
+TermListGroup::get_termfreq(const string &tname) const
+{
+    LOGCALL(API, doccount, "TermListGroup::get_termfreq", tname);
+    unordered_map<string, doccount>::const_iterator it = termfreq.find(tname);
+    if (it != termfreq.end())
+	return it->second;
+    else
+	return 0;
+}
+
+DocumentSet::DocumentSet(const DocumentSet &other)
+    : internal(other.internal)
+{
+}
+
+void
+DocumentSet::operator=(const DocumentSet &other)
+{
+    internal = other.internal;
+}
+
+DocumentSet::DocumentSet()
+    : internal(new Xapian::DocumentSet::Internal)
+{
+}
+
+unsigned int
+DocumentSet::size() const
+{
+    LOGCALL(API, int, "DocumentSet::size", NO_ARGS);
```

Comment:
Return type in the logging is `int` but the actual return type is `unsigned int`.

But actually this should probably be `Xapian::doccount` for consistency with `MSet`, etc.  Similarly for the internal equivalent.